### PR TITLE
Test for validating proxy can't be used as an open relay

### LINKED ISSUE #7697: Add an integration test that validates the proxy does not act as an open relay
What happens if a client tries to access a proxy on its outbound port (4140)?

We'd expect these connections to only be permitted from localhost. However, if the proxy and/or init/cni is misconfigured, it could be possible for an attacker to relay requests through an outbound proxy to circumvent authorization and impersonate a pod.

For example: let's consider a proxy running in ingress-mode (so that it honors the `l5d-dst-override` header). It should not be possible for an arbitrary pod to connect to the ingress-mode proxy on port:4140 so that it honors the `l5d-dst-override` header and routes requests. I would expect these attempted connections to be closed/rejected (i.e. without returning an HTTP response).

An integration test here could help us assert that changes to proxy-init/linkerd-cni or the proxy do not accidentally introduce an open relay.

### PR DESCRIPTION
**Pull request by alpeb**
Closes #7697

We setup two injected bb servers and an uninjected curl client. One of
the servers is injected in ingress mode. The curl client tries to
connect to it to the port 4140 while setting the `l5d-dst-override` header
to attempt a relay to the other server.

The first test shows this doesn't result in a relay. The second test
does the same, but changing LINKERD2_PROXY_OUTBOUND_LISTEN_ADDR in the
injected server proxy, making it listen on all addresses. In that case,
the relay succeeds.

### CODE REVIEW

File: test/integration/deep/norelay/norelay_test.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		if !strings.Contains(o, "HELLO-FROM-SERVER") {
+			testutil.AnnotatedFatalf(t, "unexpected response returned",
+				"unexpected response returned: %s", o)
+		}
+	})
+}
+
+func getDeployments(t *testing.T) map[string]string {
+	deploys := make(map[string]string)
+	var err error
```
Comment (mateiidavid) [inline test/integration/deep/norelay/norelay_test.go:125]: Pretty nitpicky but we can probably use the shorthand form on L128 and avoid declaring it explictly.

File: test/integration/deep/norelay/testdata/client.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  template:
+    metadata:
+      labels:
+        app: client
+    spec:
+      containers:
+      - name: curl
+        image: curlimages/curl
+        command: [ "sh", "-c", "--" ]
+        args: [ "while true; do sleep 10; done;" ]
```
Comment (mateiidavid) [inline test/integration/deep/norelay/testdata/client.yml:18]: Would this still work without this sleep loop?

**Review by mateiidavid — APPROVED** [review]
🙌🏻 awesome work!

File: test/integration/deep/norelay/testdata/client.yml (hunk 2 above)
Comment (alpeb) [reply test/integration/deep/norelay/testdata/client.yml:18]: No, the container will finish and k8s will continuously restart it.

**Review by adleong — APPROVED** [review]

File: test/integration/deep/norelay/testdata/client.yml (hunk 2 above)
Comment (olix0r) [inline test/integration/deep/norelay/testdata/client.yml:18]: tioli

```suggestion
        args: [ "sleep", "infinity" ]
```

**Review by kleimkuhler — APPROVED** [review]

File: test/integration/deep/norelay/norelay_test.go (hunk 1 above)
Comment (alpeb) [reply test/integration/deep/norelay/norelay_test.go:125]: You mean using this?
```go
deploys["server-hello"], err := TestHelper.LinkerdRun("inject", "testdata/server-hello.yml")
```
turns out that doesn't work when trying to set the value in a map :man_shrugging:

File: test/integration/deep/norelay/testdata/client.yml (hunk 2 above)
Comment (alpeb) [reply test/integration/deep/norelay/testdata/client.yml:18]: Oh that's deep...

Comment (alpeb) [reply test/integration/deep/norelay/testdata/client.yml]: ... but it ended up not being supported in the cURL image :man_shrugging:

**Pull request merged**